### DISCUSSION TITLE
Set Visibility After Create

### DISCUSSION THREAD

**Discussion by maldworth** [structurizr/onpremises#64] [Q&A] [answered]
I have a question that expands upon the request admin-api functionality from this issue:
https://github.com/structurizr/onpremises/issues/45

I'm attempting to automate:
1. creating a workspace (works great)
2. setting workspace role/user so it's private (because right now workspace is public by default)
3. Use the CLI to push my branches workspace.json to the on-premises install (also works great)

So is there any way I can perform step 2 through the admin api, or regular api? I tried looking at the calls that the UI makes when I set a role in the page (I'm using roles through Azure AD for my authorization), but it appears to be a form post on `.../workspace/{id}/users` so it's not an api that I can call.

I also tried looking at the DSL [visibility](https://docs.structurizr.com/dsl/language#visibility), but I don't think I want this. When developers are working in the respository and using the structurizr-lite, I don't want the dsl to have visibility = private. On the localhost it should be public. Only when I deploy to my onpremises as part of the CI/CD pipeline for the pull request branch, I would want it to create the workspace (if it doesn't exist), make it a private by default, and then use the structurizr-cli to push the workspace.json

Thank you for your time

**Comment by maldworth** [comment]
I've noticed that on this page: https://structurizr.com/help/workspace-sharing it mentions private is the default for new workspaces. I assume that's true for cloud service, but not the on-premises?

**Comment by simonbrowndotje** [comment]
`visibility` only applies to the on-premises installation and cloud service. [users](https://docs.structurizr.com/dsl/language#users) (also only applicable to the on-premises installation and cloud service) will allow you to configure who can see the workspace, which should provide a way to do (2).

**Reply by maldworth** [reply]
Thanks Simon. It should, but I don't want to list all users in my org, or use a regex. I have a role name. There is no `roles { ... }` configuration available in the DSL.

**Reply by simonbrowndotje** [reply, marked as answer]
For the on-premises installation, "username" can be a username, e-mail address, role name, or regex (https://docs.structurizr.com/onpremises/authorisation#configuring-users-and-roles). For the cloud service it's an e-mail address only.

**Reply by awehrung** [reply]
The last link seems to be dead. Is this the correct replacement? https://docs.structurizr.com/onpremises/role-based-access
